pgxn/neon/walproposer_pg.c

```diff
@@ -402,21 +405,58 @@ walprop_pg_get_shmem_state(WalProposer *wp)
 	return walprop_shared;
 }
 
-static void
-replication_feedback_set(PageserverFeedback *rf)
+/*
+ * Record new ps_feedback in the array with shards and update min_feedback.
+ */
+static PageserverFeedback
+record_pageserver_feedback(PageserverFeedback *ps_feedback)
 {
+	PageserverFeedback min_feedback;
+
+	Assert(ps_feedback->present);
+	Assert(ps_feedback->shard_number < MAX_SHARDS);
+
 	SpinLockAcquire(&walprop_shared->mutex);
-	memcpy(&walprop_shared->feedback, rf, sizeof(PageserverFeedback));
+
+	/* Update the number of shards */
+	if (ps_feedback->shard_number + 1 > walprop_shared->num_shards)
+		walprop_shared->num_shards = ps_feedback->shard_number + 1;
+
+	/* Update the feedback */
+	memcpy(&walprop_shared->shard_ps_feedback[ps_feedback->shard_number], ps_feedback, sizeof(PageserverFeedback));
+
+	/* Calculate min LSNs */
```

Comment:
Let's think about what happens when some shard hasn't sent any feedback yet: I guess that happens on a total system restart if one shard isn't available yet.

We could include the ShardCount in the pageserver feedback struct, and then refuse to calculate a valid value here unless we have heard from all the shards.  Not sure if that causes other issues.

We might want the ShardCount anyway in the feedback message: while it's relatively harmless for backpressure to be permissive in some edge case, for safekeeper updates to remote_consistent_lsn, it is important to get it exactly right (because we might drop SK data when we see remote_consistent_lsn advance), and that would require knowing the shard count.

Reply:
Backpressure before this PR (for a single pageserver) has behaved like this ([notion](https://www.notion.so/neondatabase/Backpressure-internals-cd557e16d7914e7fb51ec3a9dd29105e?pvs=4#12b045abe336452ab2d5334d8565c300)):
- after startup until receiving the first feedback, backpressure is not active (compute writes are not blocked)
- this is encoded as zero LSN (`InvalidXLogRecPtr`)

I think here it behaves the same way: if we have received feedback from a shard, it is used for min LSN calculation (and affects backpressure). If we didn't receive a feedback, then the shard is not considered for backpressure.

The other way to implement this would be to make backpressure active before receiving any feedback (block compute writes). AFAIR it was implemented this way some time ago, but was changed because it affected cold starts.

> We might want the ShardCount anyway in the feedback message

I'm not sure how useful it is. We can add it, but it would require restart of all running computes or some tricky serializations. Otherwise, old computes will spam logs with `unknown key:` messages.

Note: I don't think ShardCount is required on safekeepers and I'm not a fan of `remote_consistent_lsn` calculations on safekeeper. IMO safekeeper discovery and WAL retention can be solved with other means.

Reply:
>Note: I don't think ShardCount is required on safekeepers and I'm not a fan of remote_consistent_lsn calculations on safekeeper. IMO safekeeper discovery and WAL retention can be solved with other means.

Can you say more about the other ways of solving this problem (maybe there's an issue/design)?  Perhaps you mean we could/should do it on the broker?

Reply:
>I'm not sure how useful it is. We can add it, but it would require restart of all running computes or some tricky serializations. Otherwise, old computes will spam logs with unknown key: messages.

I didn't mean to include it unconditionally: you could do the same thing as for the ShardNumber -- only include it if it's >1

Reply:
> Can you say more about the other ways of solving this problem (maybe there's an issue/design)? 

No issue unfortunately, but the idea is:
- discovery can be done with #5691, and we also have other is_timeline_active conditions (like connected computes)
- WAL removal can use LSNs from active and past connections, and we also can have timeout before removing a segment

I think remote_consistent_lsn can be replaced with these.

Reply:
>WAL removal can use LSNs from active and past connections

Right, but we'd still need those connections to inform the safekeeper of the shardcount somehow, to know when "everyone" has seen the data (maybe you're thinking of a different place in the protocol to share that info).

Anyway -- we don't have to solve it here, I mainly wanted to make sure we weren't missing an opportunity to put the count into the structure if we would want it later.